internal/dcache/rpc/server/handler.go

```diff
@@ -47,43 +57,894 @@ var _ service.ChunkService = &ChunkServiceHandler{}
+		return nil, rpc.NewResponseError(rpc.InternalServerError, fmt.Sprintf("failed to read chunk file %s [%v]", chunkPath, err.Error()))
+	}
+
+	// get hash if requested for entire chunk
+	hash := ""
+	if req.Offset == 0 && req.Length == chunkSize {
+		hashData, err := os.ReadFile(hashPath)
+		if err != nil {
+			log.Err("ChunkServiceHandler::GetChunk: Failed to read hash file %s [%v]", hashPath, err.Error())
+			return nil, rpc.NewResponseError(rpc.InternalServerError, fmt.Sprintf("failed to read hash file %s [%v]", hashPath, err.Error()))
+		}
+		hash = string(hashData)
+	}
+
+	lmt, err := getLMT(fh)
+	if err != nil {
+		log.Err("ChunkServiceHandler::GetChunk: Failed to get LMT for chunk file %s [%v]", chunkPath, err.Error())
+		// return nil, rpc.NewResponseError(rpc.InternalServerError, fmt.Sprintf("failed to get LMT for chunk file %s [%v]", chunkPath, err.Error()))
+	}
+
+	resp := &models.GetChunkResponse{
+		Chunk: &models.Chunk{
+			Address: req.Address,
+			Data:    data,
+			Hash:    hash,
+		},
+		ChunkWriteTime: lmt,
+		TimeTaken:      time.Since(startTime).Microseconds(),
+		ComponentRV:    rvInfo.getComponentRVsForMV(req.Address.MvName),
+	}
+
+	return resp, nil
 }
 
 func (h *ChunkServiceHandler) PutChunk(ctx context.Context, req *models.PutChunkRequest) (*models.PutChunkResponse, error) {
-	return nil, nil
+	if req == nil || req.Chunk == nil || req.Chunk.Address == nil {
+		log.Err("ChunkServiceHandler::PutChunk: Received nil PutChunk request")
+		common.Assert(false, "received nil PutChunk request")
+		return nil, rpc.NewResponseError(rpc.InvalidRequest, "received nil PutChunk request")
+	}
+
+	log.Debug("ChunkServiceHandler::PutChunk: Received PutChunk request: chunk address %+v, data length %v, isSync %v", *req.Chunk.Address, req.Length, req.IsSync)
+
+	// check if the chunk address is valid
+	err := h.checkValidChunkAddress(req.Chunk.Address)
+	if err != nil {
+		log.Err("ChunkServiceHandler::PutChunk: Invalid chunk address [%s]", err.Error())
+		return nil, err
+	}
+
+	rvInfo := h.rvIDMap[req.Chunk.Address.RvID]
+	mvInfo, _ := rvInfo.getMVInfo(req.Chunk.Address.MvName)
+
+	// validate the component RVs list
+	slices.Sort(req.ComponentRV)
+	if !isComponentRVsValid(mvInfo.componentRVs, req.ComponentRV) {
+		log.Err("ChunkServiceHandler::PutChunk: Request component RVs %v are invalid for MV %s component RVs %v", req.ComponentRV, req.Chunk.Address.MvName, mvInfo.componentRVs)
+		common.Assert(false, fmt.Sprintf("invalid component RVs for MV %s", req.Chunk.Address.MvName), req.ComponentRV, mvInfo.componentRVs)
+		return nil, rpc.NewResponseError(rpc.ComponentRVsInvalid, fmt.Sprintf("request component RVs %v are invalid for MV %s component RVs %v", req.ComponentRV, req.Chunk.Address.MvName, mvInfo.componentRVs))
+	}
+
+	// check if the sync has started. If yes, block the chunk operations till the sync is completed
+	mvInfo.blockChunkOps()
```

Comment:
blockIOIfMVQuiesced()
also the comment is misleading, we do not wait till the sync completes but only for a very small time when StartSynd and EndSync are updating the sync info.